### PR TITLE
Fix various dispatcher issues

### PR DESCRIPTION
**Pull request by cmelchior**
Close https://github.com/realm/realm-kotlin/issues/1608

- This PR works around a race condition with Core notifications and dispatchers which I do not think we can solve otherwise. 
- Fixes a race condition with events in a test

See also https://github.com/Kotlin/kotlinx.coroutines/issues/3993

### CODE REVIEW

File: packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt [hunk 1]
```diff
@@ -3603,18 +3603,41 @@ actual object RealmInterop {
...
+                        // on errors instead of silent crashes.
+                        e.printStackTrace()
                     }
-                } catch (e: Exception) {
-                    // Should never happen, but is included for development to get some indicators
-                    // on errors instead of silent crashes.
-                    e.printStackTrace()
+                }
+            } catch (ex: IllegalStateException) {
+                // A work-around for https://github.com/realm/realm-kotlin/issues/1608
```
Comment (rorbech) [inline packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt:3622]: I guess we have had some debate on which way to organize the dispatcher/scheduler relationship (https://github.com/realm/realm-kotlin/pull/1559 and https://github.com/realm/realm-kotlin/pull/1463#pullrequestreview-1551509397), but this seems like a very unsafe way of guarding this. We already have infrastructure in place for being able to cancel the `SingleThreadDispatcherScheduler`, so seems way safer to expose the `cancel` to the upper layers, so that we follow the exact logic on when we expect this to happen (when we close the realm?). I guess that wouldn't work if people inject their own dispatcher and just close it, but we could document that they shouldn't close the dispatcher before closing the realm.

File: packages/test-base/src/commonTest/kotlin/io/realm/kotlin/test/common/notifications/RealmNotificationsTests.kt [hunk 2]
```diff
@@ -108,16 +108,16 @@ class RealmNotificationsTests : FlowableTests {
         realm.write { /* Do nothing */ }
         val nextVersion = realm.version()
 
+        c1.receiveOrFail(message = "Failed to receive update event on Channel 1").let { realmChange ->
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/test-base/src/commonTest/kotlin/io/realm/kotlin/test/common/notifications/RealmNotificationsTests.kt:111]: There was a race condition where the write on line 108 resulted in an update also being sent to the second observer, which would crash the test with a non-empty queue at the end. By consuming the update before listening for the 2nd initial event, this cannot happen.

File: packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt (hunk 1 above)
Comment (cmelchior) [reply packages/cinterop/src/nativeDarwin/kotlin/io/realm/kotlin/internal/interop/RealmInterop.kt]: I modified the implementation, so we now have two locks instead, one to control the dispatcher and one for the scheduler.

**Comment by cmelchior** [comment]
Ready for a second round @rorbech

File: packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt [hunk 3]
```diff
@@ -120,6 +120,10 @@ public class LiveRealmContext(
     }
 
     override fun close() {
+        // Warning: It is important to release the scheduler before closing the
```
Comment (rorbech) [inline packages/library-base/src/commonMain/kotlin/io/realm/kotlin/internal/util/CoroutineDispatcherFactory.kt:123]: I guess this will work if we are guaranteed the `user_data_free` callback from core synchronously from `realm_release(scheduler)`. It would probably be safer for future changes with a separate abstraction of the scheduler that would set `dispatcherClosing` before `realm_release(scheduler)` ... but this is way better that the previous implementation that was relying on the message from the coroutine-exception 🙈

**Review by rorbech — APPROVED** [review]
